src/data/equiv/local_equiv.lean

```diff
@@ -0,0 +1,518 @@
+/-
+Copyright (c) 2019 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import data.equiv.basic
+
+/-!
+# Local equivalences
+
+This files defines equivalences between subsets of given types.
+An element `e` of `local_equiv α β` is made of two maps `e.to_fun` and `e.inv_fun` respectively
+from α to β and from  β to α (just like equivs), which are inverse to each other on the subsets
+`e.source` and `e.target` of respectively α and β.
+
+They are designed in particular to define charts on manifolds.
+
+The main functionality is `e.trans f`, which composes the two local equivalences by restricting
+the source and target to the maximal set where the composition makes sense.
+
+Contrary to equivs, we do not register the coercion to functions and we use explicitly to_fun and
+inv_fun: coercions create numerous unification problems for manifolds.
+
+## Main definitions
```

Comment:
I like this "table" layout for `## Main definitions`. Maybe we should add advice to use it to the doc guidelines!